File: RetailShop/static/js/likes-reviews.js
/**
 * Likes and Reviews JavaScript
 * Handles dynamic likes and reviews functionality
 */

document.addEventListener('DOMContentLoaded', function() {
    // Get product ID from the page
    const productId = document.querySelector('.like-btn')?.dataset.productId;
    
    if (!productId) return;
    
    // Initialize likes
    initializeLikes(productId);
    
    // Initialize reviews
    initializeReviews(productId);
    
    // Setup like button click handler
    setupLikeButton(productId);
    
    // Setup review form submission
    setupReviewForm(productId);
    
    // Setup star rating functionality
    setupStarRating();
});

/**
 * Initialize likes for a product
 */
function initializeLikes(productId) {
    fetch(`/get_likes/${productId}`)
        .then(response => response.json())
        .then(data => {
            // Update like count
            document.getElementById(`like-count-${productId}`).textContent = data.likes_count;
            
            // Update like icon if user has liked the product
            if (data.user_liked) {
                const likeIcon = document.getElementById(`like-icon-${productId}`);
                likeIcon.classList.remove('far');
                likeIcon.classList.add('fas');
                likeIcon.style.color = '#e74c3c'; // Red color for liked
            }
        })
        .catch(error => console.error('Error fetching likes:', error));
}

/**
 * Initialize reviews for a product
 */
function initializeReviews(productId) {
    fetch(`/get_reviews/${productId}`)
        .then(response => response.json())
        .then(data => {
            // Update average rating
            document.getElementById('average-rating').textContent = data.average_rating;
            
            // Update total reviews count
            document.getElementById('total-reviews').textContent = data.total_reviews;
            document.getElementById('review-count').textContent = `(${data.total_reviews})`;
            
            // Update rating distribution
            updateRatingDistribution(data.rating_distribution, data.total_reviews);
            
            // Update average stars
            updateAverageStars(data.average_rating);
            
            // Display reviews
            displayReviews(data.reviews);
            
            // Show/hide no reviews message
            const noReviewsMessage = document.getElementById('no-reviews-message');
            if (data.reviews.length === 0) {
                noReviewsMessage.style.display = 'block';
            } else {
                noReviewsMessage.style.display = 'none';
            }
            
            // Show/hide load more button
            const loadMoreButton = document.getElementById('load-more-reviews');
            if (data.reviews.length > 5) {
                loadMoreButton.style.display = 'block';
            } else {
                loadMoreButton.style.display = 'none';
            }
        })
        .catch(error => console.error('Error fetching reviews:', error));
}

/**
 * Update rating distribution bars
 */
function updateRatingDistribution(distribution, totalReviews) {
    if (totalReviews === 0) return;
    
    // Update 5-star bar
    const fiveStarCount = distribution[5] || 0;
    const fiveStarPercentage = (fiveStarCount / totalReviews) * 100;
    document.getElementById('five-star-bar').style.width = `${fiveStarPercentage}%`;
    document.getElementById('five-star-count').textContent = fiveStarCount;
    
    // Update 4-star bar
    const fourStarCount = distribution[4] || 0;
    const fourStarPercentage = (fourStarCount / totalReviews) * 100;
    document.getElementById('four-star-bar').style.width = `${fourStarPercentage}%`;
    document.getElementById('four-star-count').textContent = fourStarCount;
    
    // Update 3-star bar
    const threeStarCount = distribution[3] || 0;
    const threeStarPercentage = (threeStarCount / totalReviews) * 100;
    document.getElementById('three-star-bar').style.width = `${threeStarPercentage}%`;
    document.getElementById('three-star-count').textContent = threeStarCount;
    
    // Update 2-star bar
    const twoStarCount = distribution[2] || 0;
    const twoStarPercentage = (twoStarCount / totalReviews) * 100;
    document.getElementById('two-star-bar').style.width = `${twoStarPercentage}%`;
    document.getElementById('two-star-count').textContent = twoStarCount;
    
    // Update 1-star bar
    const oneStarCount = distribution[1] || 0;
    const oneStarPercentage = (oneStarCount / totalReviews) * 100;
    document.getElementById('one-star-bar').style.width = `${oneStarPercentage}%`;
    document.getElementById('one-star-count').textContent = oneStarCount;
}

/**
 * Update average stars display
 */
function updateAverageStars(averageRating) {
    const starsContainer = document.getElementById('average-stars');
    const stars = starsContainer.querySelectorAll('i');
    
    // Reset all stars
    stars.forEach(star => {
        star.className = 'far fa-star';
    });
    
    // Fill stars based on average rating
    const fullStars = Math.floor(averageRating);
    const hasHalfStar = averageRating - fullStars >= 0.5;
    
    // Fill full stars
    for (let i = 0; i < fullStars; i++) {
        stars[i].className = 'fas fa-star';
    }
    
    // Add half star if needed
    if (hasHalfStar && fullStars < 5) {
        stars[fullStars].className = 'fas fa-star-half-alt';
    }
}

/**
 * Display reviews in the reviews container
 */
function displayReviews(reviews) {
    const reviewsContainer = document.getElementById('reviews-container');
    const reviewHeader = reviewsContainer.querySelector('h3');
    const loadMoreButton = document.getElementById('load-more-reviews');
    
    // Clear existing reviews (except header and load more button)
    const children = Array.from(reviewsContainer.children);
    children.forEach(child => {
        if (child !== reviewHeader && child !== loadMoreButton && child.id !== 'no-reviews-message') {
            reviewsContainer.removeChild(child);
        }
    });
    
    // Add reviews
    reviews.forEach(review => {
        const reviewCard = document.createElement('div');
        reviewCard.className = 'review-card';
        reviewCard.innerHTML = `
            <div class="review-header">
                <div class="reviewer-info">
                    <strong>${review.user_name}</strong>
                    <span class="verified-badge"><i class="fas fa-check-circle"></i></span>
                </div>
                <div class="review-stars">
                    ${getStarsHTML(review.rating)}
                </div>
            </div>
            <p class="review-text">"${review.review_text}"</p>
            <span class="review-date">Posted on ${review.created_at}</span>
        `;
        
        // Insert after the header
        reviewsContainer.insertBefore(reviewCard, reviewHeader.nextSibling);
    });
}

/**
 * Get HTML for star rating
 */
function getStarsHTML(rating) {
    let html = '';
    for (let i = 1; i <= 5; i++) {
        if (i <= rating) {
            html += '<i class="fas fa-star"></i>';
        } else {
            html += '<i class="far fa-star"></i>';
        }
    }
    return html;
}

/**
 * Setup like button click handler
 */
function setupLikeButton(productId) {
    const likeButton = document.querySelector('.like-btn');
    if (!likeButton) return;
    
    likeButton.addEventListener('click', function() {
        fetch(`/like_product/${productId}`, {
            method: 'POST',
            headers: {
                'X-Requested-With': 'XMLHttpRequest',
                'Content-Type': 'application/json'
            }
        })
        .then(response => {
            if (response.status === 401) {
                // User not logged in
                window.location.href = '/login';
                return;
            }
            return response.json();
        })
        .then(data => {
            if (!data) return;
            
            // Update like count
            document.getElementById(`like-count-${productId}`).textContent = data.likes_count;
            
            // Update like icon
            const likeIcon = document.getElementById(`like-icon-${productId}`);
            if (data.action === 'liked') {
                likeIcon.classList.remove('far');
                likeIcon.classList.add('fas');
                likeIcon.style.color = '#e74c3c'; // Red color for liked
            } else {
                likeIcon.classList.remove('fas');
                likeIcon.classList.add('far');
                likeIcon.style.color = ''; // Reset color
            }
        })
        .catch(error => console.error('Error liking product:', error));
    });
}

/**
 * Setup review form submission
 */
function setupReviewForm(productId) {
    const reviewForm = document.getElementById('review-form');
    if (!reviewForm) return;
    
    reviewForm.addEventListener('submit', function(event) {
        event.preventDefault();
        
        const rating = document.getElementById('rating-value').value;
        const reviewText = document.getElementById('review-text').value;
        
        if (!rating || rating === '0') {
            alert('Please select a rating');
            return;
        }
        
        if (!reviewText.trim()) {
            alert('Please enter a review');
            return;
        }
        
        const formData = new FormData();
        formData.append('rating', rating);
        formData.append('review_text', reviewText);
        
        fetch(`/add_review/${productId}`, {
            method: 'POST',
            headers: {
                'X-Requested-With': 'XMLHttpRequest'
            },
            body: formData
        })
        .then(response => {
            if (response.status === 401) {
                // User not logged in
                window.location.href = '/login';
                return;
            }
            return response.json();
        })
        .then(data => {
            if (!data) return;
            
            // Reset form
            reviewForm.reset();
            resetStarRating();
            
            // Refresh reviews
            initializeReviews(productId);
            
            // Show success message
            alert(data.message);
        })
        .catch(error => console.error('Error submitting review:', error));
    });
}

/**
 * Setup star rating functionality
 */
function setupStarRating() {
    const stars = document.querySelectorAll('.star-rating i');
    if (stars.length === 0) return;
    
    stars.forEach(star => {
        star.addEventListener('mouseover', function() {
            const rating = parseInt(this.dataset.rating);
            highlightStars(rating);
        });
        
        star.addEventListener('mouseout', function() {
            const currentRating = parseInt(document.getElementById('rating-value').value);
            highlightStars(currentRating);
        });
        
        star.addEventListener('click', function() {
            const rating = parseInt(this.dataset.rating);
            document.getElementById('rating-value').value = rating;
            highlightStars(rating);
        });
    });
}

/**
 * Highlight stars up to the given rating
 */
function highlightStars(rating) {
    const stars = document.querySelectorAll('.star-rating i');
    
    stars.forEach((star, index) => {
        if (index < rating) {
            star.classList.remove('far');
            star.classList.add('fas');
        } else {
            star.classList.remove('fas');
            star.classList.add('far');
        }
    });
}

/**
 * Reset star rating
 */
function resetStarRating() {
    document.getElementById('rating-value').value = '0';
    highlightStars(0);
}

File: RetailShop/likes_reviews.py
from flask import Blueprint, request, jsonify, session, redirect, url_for, flash
from RetailShop.db_helper import execute_query
import datetime

# Create a Blueprint for likes and reviews
likes_reviews = Blueprint('likes_reviews', __name__)

# Initialize database tables if they don't exist
@likes_reviews.before_app_first_request
def initialize_tables():
    # Create likes table
    execute_query("""
        CREATE TABLE IF NOT EXISTS product_likes (
            id INT AUTO_INCREMENT PRIMARY KEY,
            user_id INT NOT NULL,
            product_id INT NOT NULL,
            created_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP,
            UNIQUE KEY unique_user_product (user_id, product_id)
        )
    """, commit=True)
    
    # Create reviews table
    execute_query("""
        CREATE TABLE IF NOT EXISTS product_reviews (
            id INT AUTO_INCREMENT PRIMARY KEY,
            user_id INT NOT NULL,
            product_id INT NOT NULL,
            rating INT NOT NULL,
            review_text TEXT NOT NULL,
            created_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP,
            UNIQUE KEY unique_user_product_review (user_id, product_id)
        )
    """, commit=True)

# Route to like/unlike a product
@likes_reviews.route('/like_product/<int:product_id>', methods=['POST'])
def like_product(product_id):
    if 'uid' not in session:
        if request.headers.get('X-Requested-With') == 'XMLHttpRequest':
            return jsonify({'status': 'error', 'message': 'Please login to like products'}), 401
        flash('Please login to like products', 'danger')
        return redirect(url_for('login'))
    
    user_id = session['uid']
    
    # Check if user already liked the product
    existing_like = execute_query(
        "SELECT * FROM product_likes WHERE user_id = %s AND product_id = %s",
        (user_id, product_id),
        fetchone=True
    )
    
    if existing_like:
        # Unlike the product
        execute_query(
            "DELETE FROM product_likes WHERE user_id = %s AND product_id = %s",
            (user_id, product_id),
            commit=True
        )
        action = 'unliked'
    else:
        # Like the product
        execute_query(
            "INSERT INTO product_likes (user_id, product_id) VALUES (%s, %s)",
            (user_id, product_id),
            commit=True
        )
        action = 'liked'
    
    # Get total likes for the product
    likes_count = execute_query(
        "SELECT COUNT(*) as count FROM product_likes WHERE product_id = %s",
        (product_id,),
        fetchone=True
    )
    
    count = likes_count['count'] if likes_count else 0
    
    if request.headers.get('X-Requested-With') == 'XMLHttpRequest':
        return jsonify({
            'status': 'success',
            'action': action,
            'likes_count': count
        })
    
    # If not AJAX request, redirect back to product page
    return redirect(url_for('view_product', product_id=product_id))

# Route to get likes count for a product
@likes_reviews.route('/get_likes/<int:product_id>', methods=['GET'])
def get_likes(product_id):
    # Get total likes for the product
    likes_count = execute_query(
        "SELECT COUNT(*) as count FROM product_likes WHERE product_id = %s",
        (product_id,),
        fetchone=True
    )
    
    count = likes_count['count'] if likes_count else 0
    
    # Check if current user has liked the product
    user_liked = False
    if 'uid' in session:
        user_id = session['uid']
        user_like = execute_query(
            "SELECT * FROM product_likes WHERE user_id = %s AND product_id = %s",
            (user_id, product_id),
            fetchone=True
        )
        user_liked = user_like is not None
    
    return jsonify({
        'likes_count': count,
        'user_liked': user_liked
    })

# Route to add a review for a product
@likes_reviews.route('/add_review/<int:product_id>', methods=['POST'])
def add_review(product_id):
    if 'uid' not in session:
        if request.headers.get('X-Requested-With') == 'XMLHttpRequest':
            return jsonify({'status': 'error', 'message': 'Please login to review products'}), 401
        flash('Please login to review products', 'danger')
        return redirect(url_for('login'))
    
    user_id = session['uid']
    rating = request.form.get('rating')
    review_text = request.form.get('review_text')
    
    if not rating or not review_text:
        if request.headers.get('X-Requested-With') == 'XMLHttpRequest':
            return jsonify({'status': 'error', 'message': 'Rating and review text are required'}), 400
        flash('Rating and review text are required', 'danger')
        return redirect(url_for('view_product', product_id=product_id))
    
    # Check if user already reviewed the product
    existing_review = execute_query(
        "SELECT * FROM product_reviews WHERE user_id = %s AND product_id = %s",
        (user_id, product_id),
        fetchone=True
    )
    
    if existing_review:
        # Update existing review
        execute_query(
            "UPDATE product_reviews SET rating = %s, review_text = %s, created_at = NOW() WHERE user_id = %s AND product_id = %s",
            (rating, review_text, user_id, product_id),
            commit=True
        )
        message = 'Review updated successfully'
    else:
        # Add new review
        execute_query(
            "INSERT INTO product_reviews (user_id, product_id, rating, review_text) VALUES (%s, %s, %s, %s)",
            (user_id, product_id, rating, review_text),
            commit=True
        )
        message = 'Review added successfully'
    
    if request.headers.get('X-Requested-With') == 'XMLHttpRequest':
        # Get user info for the review
        user_info = execute_query(
            "SELECT name FROM users WHERE id = %s",
            (user_id,),
            fetchone=True
        )
        
        user_name = user_info['name'] if user_info else 'Anonymous'
        
        return jsonify({
            'status': 'success',
            'message': message,
            'review': {
                'user_name': user_name,
                'rating': int(rating),
                'review_text': review_text,
                'created_at': datetime.datetime.now().strftime("%B %d, %Y")
            }
        })
    
    flash(message, 'success')
    return redirect(url_for('view_product', product_id=product_id))

# Route to get reviews for a product
@likes_reviews.route('/get_reviews/<int:product_id>', methods=['GET'])
def get_reviews(product_id):
    # Get all reviews for the product
    reviews = execute_query(
        """
        SELECT r.*, u.name as user_name
        FROM product_reviews r
        JOIN users u ON r.user_id = u.id
        WHERE r.product_id = %s
        ORDER BY r.created_at DESC
        """,
        (product_id,),
        fetchall=True
    )
    
    # Get average rating
    avg_rating = execute_query(
        "SELECT AVG(rating) as avg_rating FROM product_reviews WHERE product_id = %s",
        (product_id,),
        fetchone=True
    )
    
    average = avg_rating['avg_rating'] if avg_rating and avg_rating['avg_rating'] else 0
    
    # Get rating distribution
    rating_distribution = execute_query(
        """
        SELECT rating, COUNT(*) as count
        FROM product_reviews
        WHERE product_id = %s
        GROUP BY rating
        ORDER BY rating DESC
        """,
        (product_id,),
        fetchall=True
    )
    
    # Format the reviews for JSON response
    formatted_reviews = []
    if reviews:
        for review in reviews:
            formatted_reviews.append({
                'id': review['id'],
                'user_name': review['user_name'],
                'rating': review['rating'],
                'review_text': review['review_text'],
                'created_at': review['created_at'].strftime("%B %d, %Y")
            })
    
    # Format rating distribution
    distribution = {1: 0, 2: 0, 3: 0, 4: 0, 5: 0}
    if rating_distribution:
        for rating in rating_distribution:
            distribution[rating['rating']] = rating['count']
    
    return jsonify({
        'reviews': formatted_reviews,
        'average_rating': round(float(average), 1) if average else 0,
        'total_reviews': len(formatted_reviews),
        'rating_distribution': distribution
    })